pydantic/types.py

```diff
@@ -191,7 +208,63 @@ class StrictStr(ConstrainedStr):
     strict = True
 
 
+class TemplateStrMeta(type):
+    # can't use __class_getattr__ since it doesnt work on python 3.6
+    def __getitem__(cls, keys: Union[str, int, Tuple[str, ...]]):
+        if isinstance(keys, int):
+            # no named keys, check their quantity
+            name = f"{cls.__name__}[{keys}]"
+            namespace = dict(__quantity__=keys)
+        else:
+            keys = set(keys) if isinstance(keys, tuple) else {keys}
+            name = f'{cls.__name__}[{", ".join(map(repr, keys))}]'
+            namespace = dict(__keys__=keys)
+
+        return new_class(name=name, bases=(cls,), exec_body=lambda ns: ns.update(namespace))
+
+
+class TemplateStr(str, metaclass=TemplateStrMeta):
+    __keys__: Set[str] = set()
+    __quantity__: int = 0
+
+    @classmethod
+    def __get_validators__(cls) -> 'CallableGenerator':
+        if cls.__keys__:
+            yield cls.validate_keys
+        else:
+            yield cls.validate_quantity
+
+    @classmethod
+    def validate_keys(cls, v) -> str:
+        v = str(v)
+        found_keys = {
+            field_name + ('!' + conversion if conversion else '') + (':' + format_spec if format_spec else '')
+            for literal_text, field_name, format_spec, conversion in Formatter().parse(v)
+            if field_name is not None
+        }
+
+        if cls.__keys__ != found_keys:
+            raise errors.TemplateStrError(cls.__keys__, found_keys)
+        return v
+
+    @classmethod
+    def validate_quantity(cls, v):
+        v = str(v)
+        expected_quantity = cls.__quantity__
+        found_keys = [
```

Comment:
Code duplication here, but decided to keep it that way for speed (otherwise list must be converted to set when we check keys)
If you have a better solution, let me know and I'll change it